File: server_node_legacy/routes/allocations.js
const express = require('express');
const db = require('../database');
const router = express.Router();

// Get all allocations
router.get('/', (req, res) => {
    db.all(`
    SELECT a.id, a.event_id, a.resource_id, e.title as event_title, r.name as resource_name, e.start_time, e.end_time
    FROM allocations a
    JOIN events e ON a.event_id = e.id
    JOIN resources r ON a.resource_id = r.id
  `, [], (err, rows) => {
        if (err) return res.status(500).json({ error: err.message });
        res.json(rows);
    });
});

// Create allocation with CONFLICT DETECTION
router.post('/', (req, res) => {
    const { event_id, resource_id } = req.body;

    // 1. Get the target event's timing
    db.get("SELECT start_time, end_time FROM events WHERE id = ?", [event_id], (err, targetEvent) => {
        if (err) return res.status(500).json({ error: err.message });
        if (!targetEvent) return res.status(404).json({ error: "Event not found" });

        const newStart = new Date(targetEvent.start_time).getTime();
        const newEnd = new Date(targetEvent.end_time).getTime();

        // 2. Check for conflicts with existing allocations for this resource
        const query = `
      SELECT e.start_time, e.end_time, e.title 
      FROM allocations a 
      JOIN events e ON a.event_id = e.id 
      WHERE a.resource_id = ?
    `;

        db.all(query, [resource_id], (err, existingAllocations) => {
            if (err) return res.status(500).json({ error: err.message });

            const conflict = existingAllocations.find(alloc => {
                const existStart = new Date(alloc.start_time).getTime();
                const existEnd = new Date(alloc.end_time).getTime();

                // Overlap logic: (StartA < EndB) && (EndA > StartB)
                return (newStart < existEnd && newEnd > existStart);
            });

            if (conflict) {
                return res.status(409).json({
                    error: "Resource conflict detected",
                    details: `Resource is already booked for '${conflict.title}' from ${conflict.start_time} to ${conflict.end_time}`
                });
            }

            // 3. No conflict, create allocation
            db.run("INSERT INTO allocations (event_id, resource_id) VALUES (?, ?)", [event_id, resource_id], function (err) {
                if (err) return res.status(500).json({ error: err.message });
                res.json({ id: this.lastID, event_id, resource_id });
            });
        });
    });
});

module.exports = router;

File: server/routes/allocations.py
from flask import Blueprint, request, jsonify
from models import db, Allocation, Event
from datetime import datetime

allocation_bp = Blueprint('allocations', __name__)

@allocation_bp.route('', methods=['GET'])
def get_allocations():
    allocations = Allocation.query.all()
    results = []
    for a in allocations:
        results.append({
            'id': a.id,
            'event_id': a.event_id,
            'resource_id': a.resource_id,
            'event_title': a.event.title,
            'resource_name': a.resource.name,
            'start_time': a.event.start_time.isoformat(),
            'end_time': a.event.end_time.isoformat()
        })
    return jsonify(results)

@allocation_bp.route('', methods=['POST'])
def create_allocation():
    data = request.get_json()
    event_id = data.get('event_id')
    resource_id = data.get('resource_id')

    # 1. Get the target event's timing
    target_event = Event.query.get(event_id)
    if not target_event:
        return jsonify({'error': 'Event not found'}), 404

    new_start = target_event.start_time
    new_end = target_event.end_time

    # 2. Check for conflicts
    # Join Allocation with Event to filter by resource_id
    existing_allocations = db.session.query(Allocation).join(Event).filter(Allocation.resource_id == resource_id).all()

    for alloc in existing_allocations:
        exist_start = alloc.event.start_time
        exist_end = alloc.event.end_time
        
        # Overlap logic: (StartA < EndB) && (EndA > StartB)
        if new_start < exist_end and new_end > exist_start:
             return jsonify({
                'error': 'Resource conflict detected',
                'details': f"Resource is already booked for '{alloc.event.title}' from {exist_start} to {exist_end}"
            }), 409

    # 3. No conflict, create
    new_alloc = Allocation(event_id=event_id, resource_id=resource_id)
    db.session.add(new_alloc)
    db.session.commit()
    return jsonify({'id': new_alloc.id, 'event_id': event_id, 'resource_id': resource_id})
